File: Cpp/AdventOfCode/src/2015/d1_SantaElevator.cpp
#include "2015/d1_SantaElevator.h"

SOLUTION(2015, 1) {
    constexpr size_t FindFloor(const std::string_view val) {
        return std::count(val.cbegin(), val.cend(), '(') - std::count(val.cbegin(), val.cend(), ')');
    }

    constexpr size_t FindBasementStep(const std::string_view val) {
        s32 currentFloor = 0;
        for (size_t step = 0; step < val.size(); step++) {
            currentFloor += (val[step] == '(' ? 1 : -1);
            if (currentFloor < 0) return step + 1;
        }
        return 0;
    }

    PART_ONE() {
        //return Constexpr::ToString(FindFloor(Line));
        return Constexpr::ToString(FindFloor(lines[0]));
    }
    PART_TWO() {
        //return Constexpr::ToString(FindBasementStep(Line));
        return Constexpr::ToString(FindBasementStep(lines[0]));
    }

    TESTS() {
        static_assert(FindFloor("") == 0);
        static_assert(FindFloor("(())") == 0);
        static_assert(FindFloor("()()") == 0);
        static_assert(FindFloor("(((") == 3);
        static_assert(FindFloor("(()(()(") == 3);
        static_assert(FindFloor("())") == -1);
        static_assert(FindFloor("))(") == -1);
        static_assert(FindFloor(")))") == -3);
        static_assert(FindFloor(")())())") == -3);

        static_assert(FindBasementStep(")") == 1);
        static_assert(FindBasementStep("()())") == 5);

        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d1_santa_elevator.rs
pub fn part_one(lines: &Vec<String>) -> String {
    find_floor(&lines[0]).to_string()
}
pub fn part_two(lines: &Vec<String>) -> String {
    find_basement_step(&lines[0]).to_string()
}

fn find_floor(val: &str) -> isize {
    let ups = val.chars().filter(|c| c == &'(').count();
    let downs = val.chars().filter(|c| c == &')').count();
    ups as isize - downs as isize
}

fn find_basement_step(val: &str) -> i32 {
    let mut current_floor: i32 = 0;
    let mut index = 0;
    for c in val.chars() {
        index += 1;
        if c == '(' {
            current_floor += 1;
        } else {
            current_floor -= 1;
            if current_floor < 0 {
                return index;
            }
        }
    }

    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn find_floor_tests() {
        assert_eq!(find_floor(""), 0);
        assert_eq!(find_floor("(())"), 0);
        assert_eq!(find_floor("()()"), 0);
        assert_eq!(find_floor("((("), 3);
        assert_eq!(find_floor("(()(()("), 3);
        assert_eq!(find_floor("())"), -1);
        assert_eq!(find_floor("))("), -1);
        assert_eq!(find_floor(")))"), -3);
        assert_eq!(find_floor(")())())"), -3);
    }

    #[test]
    fn find_basement_step_tests() {
        assert_eq!(find_basement_step(")"), 1);
        assert_eq!(find_basement_step("()())"), 5);
    }

}
